### PR TITLE
Feature: Create Note for Wiki-Article

### PR DESCRIPTION
**Pull request by muckmuck96**
- add the option to create notes for wiki articles which are directly linked to the selection/current cursor
- posibillity to choose folder where these notes should be created
- supports the template engine for the note body

### CODE REVIEW

**Review by StrangeGirlMurph — CHANGES_REQUESTED** [review]
Hey! Thank you for your PR :) The feature is great! I reviewed your code and added my thoughts.

File: src/commands/linkArticles.ts [hunk 1]
```diff
@@ -83,7 +85,45 @@ async function insert(
 	if (templateString.includes("{description}") && !article.description)
 		new Notice("The article has no description.");
 
+	if (settings.createArticleNote) {
+		if ((await createNoteForArticle(app, settings, noteTitle, insert))) {
+			insert = `[[${noteTitle}]]`;
```
[the remark this answers is not in the document]
Comment (muckmuck96) [reply src/commands/linkArticles.ts:90]: Yeah your right, I`ll add the whole path, that should work

File: src/commands/linkArticles.ts [hunk 2]
```diff
@@ -83,7 +85,45 @@ async function insert(
...
 }
+
+async function createNoteForArticle(
+	app: App, 
+	settings: WikipediaSearchSettings, 
+	title: string, 
+	content: string
+) {
+	const normalizedTitle = normalizeTitle(title);
+	const newNotePath = `${settings.createArticleNotePath}/${normalizedTitle}.md`;
```
[the remark this answers is not in the document]
Comment (muckmuck96) [reply src/commands/linkArticles.ts:106]: Thats a good point :)

File: src/settings.ts [hunk 3]
```diff
@@ -99,7 +103,30 @@ export class WikipediaSearchSettingTab extends PluginSettingTab {
...
+						.onChange(async () => {
+							settings.createArticleNotePath = text.getValue();
+							if (settings.createArticleNotePath.length == 0) {
+								settings.createArticleNotePath = DEFAULT_SETTINGS.createArticleNotePath;
+								new Notice("Empty path not available. Set to default!");
+							}
+							await this.plugin.saveSettings();
+						});
+				}
 			);
```
[the remark this answers is not in the document]
Comment (muckmuck96) [reply src/settings.ts:128]: Yeah I know this periodic-notes implementation, had some linter errros with it, but I`ll give it another try ^^

File: src/settings.ts [hunk 4]
```diff
@@ -157,6 +184,22 @@ export class WikipediaSearchSettingTab extends PluginSettingTab {
...
+					if (settings.createArticleNote) {
+						settings.defaultTemplate = "{title}\n{thumbnail}\n{intro}";
+					} else {
+						settings.defaultTemplate = DEFAULT_SETTINGS.defaultTemplate;
+					}
+					await this.plugin.saveSettings();
+					this.display();
+				})
+		);
+
```
[the remark this answers is not in the document]
Comment (muckmuck96) [reply src/settings.ts:202]: Thats a good idea, I`ll transport this to the additional templates :)

**Comment by muckmuck96** [comment]
Ok I have adjusted the whole thing so that the configuration can now be set per Additional Template. Also I made a dropdown from the path search for the storage folder.

File: src/commands/linkArticles.ts [hunk 5]
```diff
@@ -83,7 +87,39 @@ async function insert(
 	if (templateString.includes("{description}") && !article.description)
 		new Notice("The article has no description.");
 
+	if (template.createArticleNote) {
+		const createdNote = await createNoteInFolder(app, noteTitle, insert, template);
+		if (createdNote != null) {
+			insert = `[${noteTitle}](${createdNote})`;
```
Comment (StrangeGirlMurph) [inline src/commands/linkArticles.ts:93]: This is not how linking to notes with a path works in Obsidian. It's `[[<path>|<name>]]` (Try it out by creating two files with the same name and reference both. Then view the links in the src view). 
This also should match the new link format setting 
image

**Review by StrangeGirlMurph — CHANGES_REQUESTED** [review]
This is just what I noticed at a glance. I'll take a look at how you made the settings when I find the time. Please also go through my old comments and mark them as resolved if they are. Also I hope you don't forget to test your feature thoroughly for edge cases and stuff :)

File: src/commands/linkArticles.ts (hunk 5 above)
Comment (muckmuck96) [reply src/commands/linkArticles.ts:93]: yeah, thats my fault ^^ misplaced that with standard markdown.

**Comment by muckmuck96** [comment]
Yes, I had overlooked that. ^^ I had already tested some edge cases and that looked so far quite good. This morning I had noticed something unround in the settings, I have still supplemented. I think now could be only small things in the settings, which you might not quite agree with the user experience. I for my part, however, find that already quite successful. :)

File: src/commands/linkArticles.ts [hunk 6]
```diff
@@ -83,7 +86,42 @@ async function insert(
...
+		}
+		const createdNote = await createNoteInFolder(app, noteTitle, insert, template);
+		if (createdNote != null) {
+			insert = `[[${createdNote}|${noteTitle}]]`;
+		}
+	}
+
 	const cursorPosition = editor.getCursor();
 	editor.replaceSelection(insert);
 	if (settings.placeCursorInfrontOfInsert) editor.setCursor(cursorPosition);
```
Comment (StrangeGirlMurph) [inline src/commands/linkArticles.ts:101]: If the note couldn't be created nothing should happen. Don't just insert the content that is meant to be in a separate file directly inside the current file.

File: src/commands/linkArticles.ts [hunk 7]
```diff
@@ -83,7 +86,42 @@ async function insert(
...
+	app: App,
+	title: string,
+	content: string,
+	template: Template
+): Promise<string | null> {
+	const newNotePath = normalizePath(`${template.createArticleNoteCustomPath}/${title}.md`);
+	const existingFile = app.vault.getAbstractFileByPath(newNotePath);
+	
+	if (existingFile) {
+		new Notice(`File already exists in the current folder. Instead directly linked article on cursor/selection.`);
```
Comment (StrangeGirlMurph) [inline src/commands/linkArticles.ts:114]: This notice should match my previous request regarding lines 89 to 101. Also drop the "current"

File: src/settings.ts [hunk 8]
```diff
@@ -103,6 +118,29 @@ export class WikipediaSearchSettingTab extends PluginSettingTab {
...
+				.setName("Default path for created notes")
+				.setDesc("Folder where created notes should be saved. Set empty to use the vault root folder. (type to search)")
+					.addDropdown((dropdown: DropdownComponent) => {
+						dropdown
+							.addOptions(this.app.vault.getAllLoadedFiles()
+								.filter(f => (f instanceof TFolder))
+								.reduce((acc, item) => {
+									acc[item.name] = item.name;
+									return acc;
+								}, {} as Record<string, string>))
```
Comment (StrangeGirlMurph) [inline src/settings.ts:132]: This isn't a good solution. Directories with the same name aren't distinguishable...

File: src/settings.ts [hunk 9]
```diff
@@ -135,7 +178,64 @@ export class WikipediaSearchSettingTab extends PluginSettingTab {
...
+							.onChange(async (value) => {
+								settings.templates[i].createArticleNote = value;
+								if (settings.templates[i].createArticleNote) {
+									settings.templates[i].templateString = "{thumbnail}\n{intro}\n\n{url}";
+								} else {
+									settings.templates[i].templateString = DEFAULT_SETTINGS.defaultTemplate;
+								}
+								await this.plugin.saveSettings();
+								this.display();
+							})
```
Comment (StrangeGirlMurph) [inline src/settings.ts:198]: I don't like the idea that it would just instantly override any previous template. Maybe only change the template to `{thumbnail}\n{intro}\n\n{url}` or something if the old template is the default one.

**Review by StrangeGirlMurph — CHANGES_REQUESTED** [review]
There are still some big problems. I am not really happy with how the settings turned out. I think they are pretty confusing. There is a lot of repetition with the descriptions. The design is something I can work out myself though.

File: src/commands/linkArticles.ts (hunk 7 above)
Comment (muckmuck96) [reply src/commands/linkArticles.ts:114]: You mean the one with the normalizedTitle variable in the message?

File: src/settings.ts (hunk 9 above)
Comment (muckmuck96) [reply src/settings.ts:198]: should I then leave it in so that when I turn off the Create Note function, the template when it was set to the one I still set back to the default template or should that then just stay as it is?

File: src/commands/linkArticles.ts (hunk 7 above)
Comment (StrangeGirlMurph) [reply src/commands/linkArticles.ts:114]: I don't understand. I mean the one I highlighted in line 114. My other request says that if no note could be created the plugin should do nothing instead of bringing the content that was meant for the note into the current file. If this behavior is implemented this notice needs to reflect that. That's what I mean. Do you understand?

File: src/settings.ts (hunk 9 above)
Comment (StrangeGirlMurph) [reply src/settings.ts:198]: Stay as is please :)

File: src/commands/linkArticles.ts (hunk 7 above)
Comment (muckmuck96) [reply src/commands/linkArticles.ts:114]: Ok now I understood, what you meant by your previous reqeust. Yeah I`ll change that as well.

File: src/settings.ts (hunk 8 above)
Comment (muckmuck96) [reply src/settings.ts:132]: Yeah thats true. I tried to use the path as a display element for the option, but the dropdown element then increases immeasurably the deeper you go into a folder branch. That's actually a shame. I've seen the FolderSuggest from Periodic Notes many times and used it myself in other plugins, but it has additional dependencies and sometimes linting errors. What would be your approach?

Comment (muckmuck96) [reply src/settings.ts:132]: Otherwise I´ll just add the FolderSuggester...

**Review by StrangeGirlMurph — COMMENTED** [review]
Thank you for this feature and your help in general :blush: I'll take over and make things pretty now. I hope you don't mind me doing this directly here.

**Comment by muckmuck96** [comment]
When are you planning a new update for the feature? I have a few interested people waiting for this feature :)

**Comment by StrangeGirlMurph** [comment]
@muckmuck96 I am nearly done! I just have to make it a bit more responsible for mobile and have to decide on whether or not I want to do data migration. This currently includes some breaking changes that would occur when loading the old settings data. I'll hopefully be done by the end of the day.

**Review by StrangeGirlMurph — APPROVED** [review]
All finished now I think :)

**Comment by ValiHar** [comment]
I love the feature! Thanks for implementing it so fast :) We could further enhance this by adding support for templater templates 😁 What do you thing @muckmuck96 & @StrangeGirlMurph

**Comment by StrangeGirlMurph** [comment]
@ValiHar great idea! Thanks for sharing. I created an issue for this. Please have a look and go into a bit more detail on what you had in mind :)
